README
Python-apt is a wrapper to use features of apt from python.

It contains the following modules:

C++ Wrapper:
------------
apt_pkg - access to libapt-pkg (wrapper to the lowlevel c++ code)
apt_inst - access to libapt-inst (wrapper to the lowlevel c++ code)

Python module:
--------------
WARNING: the API of this module is not stable yet

apt - high level python interface build on top of apt_pkg, apt_inst


Running the tests:
------------------

Run: 
python tests/test_all.py


Running mypy:
-------------

To check if the "apt" python module is mypy clean, run:
$ MYPYPATH=./typehinting/ mypy ./apt

To use the annotation with your source code, run:
$ mypy ./my-program

debian/changelog
python-apt (2.2.1) unstable; urgency=medium

  [ Debian Janitor ]
  * Apply multi-arch hints
    - python-apt-common: Add Multi-Arch: foreign (Closes: #968458)

  [ Julian Andres Klode ]
  * Updated mirror lists (and mirror list location for Debian)

 -- Julian Andres Klode <jak@debian.org>  Sun, 04 Jul 2021 13:57:51 +0200

python-apt (2.2.0) unstable; urgency=medium

  [ Helmut Grohne ]
  * Annotate test dependencies with <!nocheck> (Closes: #980719)

  [ Julian Andres Klode ]
  * debian/rules: Re-enable PIE hardening
  * Mirror list updates
  * debfile: Pass `Name` instead of `Binary` to ExtractTar.
    Passing the binary causes it to fail trying to find the `false`
    compressor when a binary for a given compressor is not installed.
    (LP: #1926437)
  * debian/gbp.conf: Branch name is main

 -- Julian Andres Klode <jak@debian.org>  Thu, 29 Apr 2021 09:49:08 +0200

python-apt (2.1.7) unstable; urgency=medium

  * SECURITY UPDATE: various memory and file descriptor leaks (LP: #1899193)
    - python/arfile.cc, python/generic.h, python/tag.cc, python/tarfile.cc:
      fix file descriptor and memory leaks
    - python/apt_instmodule.cc, python/apt_instmodule.h, python/arfile.h:
      Avoid reference cycle with control,data members in apt_inst.DebFile
      objects
    - tests/test_cve_2020_27351.py: Test cases for DebFile (others not easily
      testable)
  * Regression fixes for the updates merged too:
    - arfile.cc: Fix segmentation fault when opening fd, track lifetime correctly
      (Closes: #977000)
    - arfile: Regression: Collect file<->deb/ar reference cycles

 -- Julian Andres Klode <jak@debian.org>  Thu, 10 Dec 2020 15:35:32 +0100

python-apt (2.1.6) unstable; urgency=medium

  [ Julian Andres Klode ]
  * Rename Cdrom and Disc to Installation medium (LP: #1849406)
  * hashes: Release GIL while hashing stuff

  [ Niels Thykier ]
  * Correct error message in hashes.cc

 -- Julian Andres Klode <jak@debian.org>  Wed, 25 Nov 2020 12:47:37 +0100

python-apt (2.1.5) unstable; urgency=medium

  * Do not require hardcoding release names in distro info, cleanup Debian and Ubuntu
    (LP: #1727470)
  * Debian: use -security instead of /updates for >= 11

 -- Julian Andres Klode <jak@debian.org>  Tue, 27 Oct 2020 21:27:56 +0100

python-apt (2.1.4) unstable; urgency=medium

  [ Debian Janitor ]
  * Trim trailing whitespace.
  * Wrap long lines in changelog entries: 1.9.9, 1.7.0~alpha1,
    1.1.0~alpha1, 0.9.4, 0.9.3.4.
  * Update standards version to 4.5.0, no changes needed.

  [ Dave Jones ]
  * Fix ftbfs with sphinx 3.x (Closes: #963655)

  [ Julian Andres Klode ]
  * Make pycodestyle happy
  * Pacify mypy 0.790 for NotImplemented returns
  * gitlab-ci: Pull debian:unstable, not ubuntu:focal

 -- Julian Andres Klode <jak@debian.org>  Thu, 15 Oct 2020 14:25:06 +0200

python-apt (2.1.3) unstable; urgency=medium

  * data/templates/Ubuntu.info.in: Add groovy (LP: #1874880)

 -- Julian Andres Klode <jak@debian.org>  Sat, 25 Apr 2020 13:05:50 +0200

python-apt (2.1.2) unstable; urgency=medium

  * Use pybuild to split out -dbg stuff, avoid .install/.files (Closes: #958118)
  * Remove Python 2 and old pyflakes hacks

 -- Julian Andres Klode <jak@debian.org>  Mon, 20 Apr 2020 18:16:28 +0200

python-apt (2.1.1) unstable; urgency=medium

  * Make python-apt-dev depend on python3-apt, not python-apt

 -- Julian Andres Klode <jak@debian.org>  Thu, 16 Apr 2020 09:54:42 +0200

python-apt (2.1.0) unstable; urgency=medium

  [ Gordon Ball ]
  * Use pyflakes3 instead of (python 2) pyflakes

  [ Julian Andres Klode ]
  * Stop building for Python 2 (Closes: #937579)
  * Remove Python 2 autopkgtests

 -- Julian Andres Klode <jak@debian.org>  Thu, 16 Apr 2020 09:46:49 +0200

python-apt (2.0.0) unstable; urgency=medium

  [ Valters Jansons ]
  * aptsources: Allow non-existent `source.list` file

  [ Julian Andres Klode ]
  * Workaround mypy 0.770
  * whatsnew: Move 1.9 development series to 2.0
  * Update mirror lists
  * Add redundant python{,3}-all build depends for lintian

  [ Simon Poirier ]
  * Decode locale-encoded descriptions.

 -- Julian Andres Klode <jak@debian.org>  Thu, 09 Apr 2020 09:16:18 +0200

python-apt (1.9.10) unstable; urgency=medium

  * Improve type annotations for apt_pkg.TagFile and apt_pkg.TagSection.
    Thanks to Stuart Prescott for the report that object was ... wrong
  * Break broken kthresher and apt-xapian-index versions

 -- Julian Andres Klode <jak@debian.org>  Sun, 08 Mar 2020 15:16:44 +0100

python-apt (1.9.9) unstable; urgency=medium

  [ Julian Andres Klode ]
  * Upload to unstable

  [ Dan Streetman ]
  * distinfo: correct subprocess usage
  * sourceslist: in __find() rstrip / from uri when comparing
  * test: add test to verify identical entries except trailing slash are
    considered equal

 -- Julian Andres Klode <jak@debian.org>  Sat, 07 Mar 2020 21:26:04 +0100

python-apt (1.9.8) experimental; urgency=medium

  * Fix build against apt 1.9.11 (Index -> MapPointer)

 -- Julian Andres Klode <jak@debian.org>  Wed, 26 Feb 2020 21:32:54 +0100

python-apt (1.9.7) experimental; urgency=medium

  * debian/tests/run-tests: Use py{3,}versions -s, not py{,3}versions -r;
    fixes regression in autopkgtest due to stderr output.

 -- Julian Andres Klode <jak@debian.org>  Sat, 22 Feb 2020 17:51:04 +0100

python-apt (1.9.6) experimental; urgency=medium

  [ dann frazier ]
  * Drop remaining references to pep8

  [ Dan Streetman ]
  * when comparing SourceEntries, rstrip '/' from uris

  [ Julian Andres Klode ]
  * gitlab-ci: pull from focal
  * Use debhelper-compat (= 12) instead of debian/compat 9
  * Use pybuild for building
  * gitlab-ci.yml: Publish html doc build by package, do not rebuild
  * Correctly install dbg variants of modules for python 3.8
  * gitlab-ci: Add daily built apt PPA
  * Switch from removed Summation classes to Hashes class
  * policy: Add set_priority(PackageFile, int)
  * policy: If no candidate version could be found, return None.

setup3.py
#! /usr/bin/env python3
# $Id: setup.py,v 1.2 2002/01/08 07:13:21 jgg Exp $
import glob
import os
import shutil
import sys

from distutils.core import setup, Extension
from distutils.sysconfig import parse_makefile
#from DistUtilsExtra.command import build_extra, build_i18n


# The apt_pkg module
files = ["python/"+source for source in parse_makefile("python/makefile")["APT_PKG_SRC"].split()]
apt_pkg = Extension("apt_pkg", files, libraries=["apt-pkg"])

# The apt_inst module
files = ["python/"+source for source in parse_makefile("python/makefile")["APT_INST_SRC"].split()]
apt_inst = Extension("apt_inst", files, libraries=["apt-pkg", "apt-inst"])

# Replace the leading _ that is used in the templates for translation
templates = []

# build doc
if len(sys.argv) > 1 and sys.argv[1] == "build":
    if not os.path.exists("build/data/templates/"):
        os.makedirs("build/data/templates")
    for template in glob.glob('data/templates/*.info.in'):
        source = open(template, "r")
        build = open(os.path.join("build", template[:-3]), "w")
        lines = source.readlines()
        for line in lines:
            build.write(line.lstrip("_"))
        source.close()
        build.close()


if len(sys.argv) > 1 and sys.argv[1] == "clean" and '-a' in sys.argv:
    for dirname in "build/doc", "doc/build", "build/data", "build/mo":
        if os.path.exists(dirname):
            print("Removing", dirname)
            shutil.rmtree(dirname)
        else:
            print("Not removing", dirname, "because it does not exist")

setup(name="python-apt",
      description="Python bindings for APT",
      version=os.environ.get('DEBVER'),
      author="APT Development Team",
      author_email="deity@lists.debian.org",
      ext_modules=[apt_pkg, apt_inst],
      packages=['apt', 'apt.progress', 'aptsources'],
      data_files = [('share/python-apt/templates',
                    glob.glob('build/data/templates/*.info')),
                    ('share/python-apt/templates',
                    glob.glob('data/templates/*.mirrors'))],
#      cmdclass = {"build": build_extra.build_extra,
#                  "build_i18n": build_i18n.build_i18n},
      license = 'GNU GPL',
      platforms = 'posix')

if len(sys.argv) > 1 and sys.argv[1] == "build":
    try:
        import sphinx
    except ImportError:
        print(('W: Sphinx not available - Not building'
                               'documentation'), file=sys.stderr)
    try:
        import pygtk
    except ImportError:
        print(('W: Not building documentation because python-'
                              'gtk2 is not available at the moment.'), file=sys.stderr)
        sys.exit(0)
    sphinx.main(["sphinx", "-b", "html", "-d", "build/doc/doctrees",
                os.path.abspath("doc/source"), "build/doc/html"])
    sphinx.main(["sphinx", "-b", "text", "-d", "build/doc/doctrees",
                os.path.abspath("doc/source"), "build/doc/text"])
